syntax = "proto3";
package remote_torch;

message Reference {
    string identifier = 1;
    string description = 2;
}

message Chunk {
    bytes data = 1;
    string description = 2;
    bytes secret = 3;
}

message Empty {}

message TrainConfig {
    Reference model = 1;
    Reference dataset = 2;
    int32 batch_size = 3;
    int32 epochs = 4;
    float learning_rate = 5;
}

message PredictConfig {
    Reference model = 1;
    Reference dataset = 2;
    int32 batch_size = 3;
}

message References {
    repeated Reference list = 1;
}  

message Accuracy {
    float value = 1;
}

service RemoteTorch {
    rpc SendDataset (stream Chunk) returns (Reference) {}
    rpc SendModel (stream Chunk) returns (Reference) {}
    rpc FetchDataset (Reference) returns (stream Chunk) {}
    rpc FetchModule (Reference) returns (stream Chunk) {}
    rpc DeleteDataset (Reference) returns (Empty) {}
    rpc DeleteModule (Reference) returns (Empty) {}
    rpc AvailableModels(Empty) returns (References) {}
    rpc AvailableDatasets(Empty) returns (References) {}
    rpc Train (TrainConfig) returns (Reference) {}
    rpc Predict (PredictConfig) returns (Reference) {}
    rpc Test (PredictConfig) returns (Accuracy) {}
}
